File: backend/apps/api/src/main/java/com/ayur/admin/web/rest/AuthController.java
package com.ayur.admin.web.rest;

import com.ayur.admin.service.AuthService;
import com.ayur.admin.service.dto.LoginRequest;
import com.ayur.admin.service.dto.LoginResponse;
import com.ayur.admin.service.dto.TwoFaEnableResponse;
import com.ayur.admin.service.dto.TwoFaVerifyRequest;
import com.ayur.admin.service.dto.UserProfileResponse;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/auth")
@Tag(name = "Authentication", description = "Authentication and user management endpoints")
@RequiredArgsConstructor
public class AuthController {

  private final AuthService authService;

  @PostMapping("/login")
  @Operation(summary = "User login", description = "Authenticate user with username and password")
  public ResponseEntity<LoginResponse> login(@Valid @RequestBody LoginRequest request) {
    LoginResponse response = authService.login(request);
    return ResponseEntity.ok(response);
  }

  @PostMapping("/refresh")
  @Operation(
      summary = "Refresh access token",
      description = "Get new access token using refresh token")
  public ResponseEntity<LoginResponse> refresh(
      @RequestHeader("X-Refresh-Token") String refreshToken) {
    LoginResponse response = authService.refreshToken(refreshToken);
    return ResponseEntity.ok(response);
  }

  @PostMapping("/logout")
  @Operation(summary = "User logout", description = "Invalidate current user session")
  public ResponseEntity<Void> logout(Authentication authentication) {
    authService.logout(authentication.getName());
    return ResponseEntity.noContent().build();
  }

  @GetMapping("/me")
  @Operation(
      summary = "Get current user",
      description = "Retrieve current authenticated user profile")
  public ResponseEntity<UserProfileResponse> getCurrentUser(Authentication authentication) {
    UserProfileResponse response = authService.getCurrentUserProfile(authentication.getName());
    return ResponseEntity.ok(response);
  }

  @PostMapping("/2fa/enable")
  @PreAuthorize("isAuthenticated()")
  @Operation(
      summary = "Enable 2FA",
      description = "Enable two-factor authentication for current user")
  public ResponseEntity<TwoFaEnableResponse> enableTwoFa(Authentication authentication) {
    TwoFaEnableResponse response = authService.enableTwoFa(authentication.getName());
    return ResponseEntity.ok(response);
  }

  @PostMapping("/2fa/verify")
  @PreAuthorize("isAuthenticated()")
  @Operation(
      summary = "Verify 2FA code",
      description = "Verify TOTP code to complete 2FA setup or login")
  public ResponseEntity<Void> verifyTwoFa(
      @Valid @RequestBody TwoFaVerifyRequest request, Authentication authentication) {
    authService.verifyTwoFa(authentication.getName(), request.getCode());
    return ResponseEntity.ok().build();
  }

  @DeleteMapping("/2fa/disable")
  @PreAuthorize("isAuthenticated()")
  @Operation(
      summary = "Disable 2FA",
      description = "Disable two-factor authentication for current user")
  public ResponseEntity<Void> disableTwoFa(Authentication authentication) {
    authService.disableTwoFa(authentication.getName());
    return ResponseEntity.noContent().build();
  }
}

File: ayurveda-api/src/auth/auth.controller.ts
import {
  Controller,
  Post,
  Get,
  Delete,
  Body,
  UseGuards,
  Headers,
} from '@nestjs/common';
import { ApiTags, ApiOperation, ApiBearerAuth } from '@nestjs/swagger';
import { AuthService } from './auth.service';
import { LoginDto } from './dto/login.dto';
import { RegisterDto } from './dto/register.dto';
import { JwtAuthGuard } from './guards/jwt-auth.guard';
import { CurrentUser } from '../common/decorators/current-user.decorator';
import { Public } from '../common/decorators/public.decorator';

@ApiTags('Authentication')
@Controller('api/auth')
export class AuthController {
  constructor(private authService: AuthService) {}

  @Public()
  @Post('register')
  @ApiOperation({ summary: 'Register new user' })
  register(@Body() dto: RegisterDto) {
    return this.authService.register(dto);
  }

  @Public()
  @Post('login')
  @ApiOperation({ summary: 'Login user' })
  login(@Body() dto: LoginDto) {
    return this.authService.login(dto);
  }

  @Public()
  @Post('refresh')
  @ApiOperation({ summary: 'Refresh access token' })
  refreshToken(@Headers('x-refresh-token') refreshToken: string) {
    return this.authService.refreshToken(refreshToken);
  }

  @UseGuards(JwtAuthGuard)
  @Post('logout')
  @ApiBearerAuth()
  @ApiOperation({ summary: 'Logout user' })
  logout(@CurrentUser('id') userId: string) {
    return this.authService.logout(userId);
  }

  @UseGuards(JwtAuthGuard)
  @Get('me')
  @ApiBearerAuth()
  @ApiOperation({ summary: 'Get current user profile' })
  getCurrentUser(@CurrentUser('id') userId: string) {
    return this.authService.getCurrentUserProfile(userId);
  }

  @UseGuards(JwtAuthGuard)
  @Post('2fa/enable')
  @ApiBearerAuth()
  @ApiOperation({ summary: 'Enable 2FA' })
  enableTwoFa(@CurrentUser('id') userId: string) {
    return this.authService.enableTwoFa(userId);
  }

  @UseGuards(JwtAuthGuard)
  @Post('2fa/verify')
  @ApiBearerAuth()
  @ApiOperation({ summary: 'Verify 2FA code' })
  verifyTwoFa(
    @CurrentUser('id') userId: string,
    @Body('code') code: string,
  ) {
    return this.authService.verifyTwoFa(userId, code);
  }

  @UseGuards(JwtAuthGuard)
  @Delete('2fa/disable')
  @ApiBearerAuth()
  @ApiOperation({ summary: 'Disable 2FA' })
  disableTwoFa(@CurrentUser('id') userId: string) {
    return this.authService.disableTwoFa(userId);
  }
}
